### PR TITLE
Mimecast search command - do not merge yet

### PR DESCRIPTION
**Pull request by yardensade**
This is pending test on client's env since we don't have a working instance

### CODE REVIEW

**Comment by yardensade** [comment]
https://github.com/demisto/etc/issues/19126

File: Integrations/MimecastV2/MimecastV2.py [hunk 1]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+        advanced_params['senderIP'] = sender_ip
+
+    search_params['advancedTrackAndTraceOptions'] = advanced_params
+    tracked_mails = search_mail_request(search_params)
+    if limit:
+        tracked_mails = tracked_mails[:limit]
+
+    for tracked_mail in tracked_mails:
+        to = []  # type: List[Any]
+        receivers = tracked_mail.get('to')
```
Comment (yardensade) [inline Integrations/MimecastV2/MimecastV2.py]:
```suggestion
        receivers = tracked_mail.get('to', [])
```

File: Integrations/MimecastV2/MimecastV2.py [hunk 2]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+    tracked_mails = search_mail_request(search_params)
+    if limit:
+        tracked_mails = tracked_mails[:limit]
+
+    for tracked_mail in tracked_mails:
+        to = []  # type: List[Any]
+        receivers = tracked_mail.get('to', [])
+        for receiver in receivers:
+            to.append({
+                'RecipientAddress': receiver.get('recipientAddress'),
```
Comment (yardensade) [inline Integrations/MimecastV2/MimecastV2.py]:
```suggestion
                'DisplayableName': receiver.get('displayableName'),
```

File: Integrations/MimecastV2/MimecastV2.py [hunk 3]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+                'TaggedExternal': receiver.get('taggedExternal')
```
Comment (yardensade) [inline Integrations/MimecastV2/MimecastV2.py]:
```suggestion
                'EmailAddress': receiver.get('emailAddress')
```

**Comment by kirbles19** [comment]
@yardensade Done. Will need to do a full review of the integration for next release.

File: Integrations/MimecastV2/MimecastV2.yml [hunk 4]
```diff
@@ -32,8 +32,8 @@ configuration:
   name: secretKey
   required: false
   type: 4
-- defaultvalue: 'true'
-  display: Trust any certificate (unsecure)
+- defaultvalue: ''
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.yml]: revert

File: Integrations/MimecastV2/MimecastV2.py [hunk 5]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+    advanced_params = {}  # type: Dict[Any, Any]
+    search_reason = demisto.args().get('search_reason', '').encode('utf-8')
+    start = demisto.args().get('start', '').encode('utf-8')
+    end = demisto.args().get('end', '').encode('utf-8')
+    sender_ip = demisto.args().get('sender_ip', '').encode('utf-8')
+    mail_to = demisto.args().get('mail_to', '').encode('utf-8')
+    mail_from_ = demisto.args().get('mail_from', '').encode('utf-8')
+    subject = demisto.args().get('subject', '').encode('utf-8')
+    message_id = demisto.args().get('message_id', '').encode('utf-8')
+    limit = int(demisto.args().get('limit', 100))
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]: add limit to the arguments with default value 100

File: Integrations/MimecastV2/MimecastV2.py [hunk 6]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+    contents = []  # type: List[Any]
+    context = {}  # type: Dict[Any, Any]
+    search_params = {}  # type: Dict[Any, Any]
+    advanced_params = {}  # type: Dict[Any, Any]
+    search_reason = demisto.args().get('search_reason', '').encode('utf-8')
+    start = demisto.args().get('start', '').encode('utf-8')
+    end = demisto.args().get('end', '').encode('utf-8')
+    sender_ip = demisto.args().get('sender_ip', '').encode('utf-8')
+    mail_to = demisto.args().get('mail_to', '').encode('utf-8')
+    mail_from_ = demisto.args().get('mail_from', '').encode('utf-8')
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]:
```suggestion
    mail_from = demisto.args().get('mail_from', '').encode('utf-8')
```

File: Integrations/MimecastV2/MimecastV2.py [hunk 7]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+def search_mail_request(search_params):
+    # Setup required variables
+    api_endpoint = '/api/message-finder/search'
+    payload = {
+        'data': [search_params]
+    }
+
+    response = http_request('POST', api_endpoint, str(payload))
+    if response.get('fail'):
+        return_error(json.dumps(response.get('fail')[0].get('errors')))
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]: lets verify that the output of `json.dumps(response.get('fail')` is a list is not empty and that the output is a dict

File: Integrations/MimecastV2/MimecastV2.py [hunk 8]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+    return response.get('data')[0].get('trackedEmails')
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]: lets verify that the output of `response.get('data')` is a list is not empty and that the output is a dict

File: Integrations/MimecastV2/MimecastV2.py [hunk 9]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+        advanced_params['to'] = mail_to
+    if subject:
+        advanced_params['subject'] = subject
+    if sender_ip:
+        advanced_params['senderIP'] = sender_ip
+
+    search_params['advancedTrackAndTraceOptions'] = advanced_params
+    tracked_mails = search_mail_request(search_params)
+    if limit:
+        tracked_mails = tracked_mails[:limit]
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]: lets verify `tracked_mails` is list

File: Integrations/MimecastV2/MimecastV2.py [hunk 10]
```diff
@@ -1479,6 +1479,101 @@ def download_attachment_request(attachment_id):
...
+        })
+
+    context['Mimecast.TrackedEmails'] = contents
+
+    results = {
+        'Type': entryTypes['note'],
+        'ContentsFormat': formats['json'],
+        'Contents': contents,
+        'ReadableContentsFormat': formats['markdown'],
+        'HumanReadable': tableToMarkdown('Mimecast Search: ', contents, headers),
```
Comment (Itay4) [inline Integrations/MimecastV2/MimecastV2.py]:
```suggestion
        'HumanReadable': tableToMarkdown('Mimecast Search', contents, headers),
```

File: Integrations/MimecastV2/MimecastV2.yml (hunk 4 above)
Comment (yardensade) [reply Integrations/MimecastV2/MimecastV2.yml]: are we still reverting this? if so we need to cancel the hook + tell everyone who changed the default value according to hooks to revert as well.

File: Integrations/MimecastV2/MimecastV2.py (hunk 9 above)
Comment (yardensade) [reply Integrations/MimecastV2/MimecastV2.py]: Now the tracked emails will be list also if nothing returns, so no need to revert

**Comment by only-says-nice** [comment]
Nice

**Pull request closed without merging**